File: java/FindCommonCharacters1002.java
import java.util.ArrayList;
import java.util.List;

/*
Given a string array words, return an array of all characters that show up in all strings within the words (including duplicates). You may return the answer in any order.
*/
public class FindCommonCharacters1002 {
    public static void main(String[] args) {
        String[] words = {"bella", "label", "roller"};
        System.out.println(commonChars(words));

        String[] words2 = {"cool", "lock", "cook"};
        System.out.println(commonChars(words2));

        String[] words3 = {"bella", "label", "roller"};
        System.out.println(commonChars(words3));
    }

    public static List<String> commonChars(String[] words) {
        List<String> result = new ArrayList<>();
        int[] commonCharacterCounts = new int[26];
        for (char c : words[0].toCharArray()) commonCharacterCounts[c - 'a']++;
        for (int i = 1; i < words.length; i++) {
            int[] currentCharacterCounts = new int[26];
            for (char c : words[i].toCharArray()) currentCharacterCounts[c - 'a']++;
            for (int j = 0; j < 26; j++)
                commonCharacterCounts[j] = Math.min(commonCharacterCounts[j], currentCharacterCounts[j]);
        }
        for (int i = 0; i < 26; i++)
            for (int j = 0; j < commonCharacterCounts[i]; j++) result.add(Character.toString(i + 'a'));
        return result;
    }
}

File: python/FindCommonCharacters1002.py
# Given a string array words, return an array of all characters that show up in all strings within the words (
# including duplicates). You may return the answer in any order.
from typing import List


def commonChars(words: List[str]) -> List[str]:
    common_character_counts = [0] * 26
    for c in words[0]:
        common_character_counts[ord(c) - ord('a')] += 1
    for word in words[1:]:
        current_character_counts = [0] * 26
        for c in word:
            current_character_counts[ord(c) - ord('a')] += 1
        for i in range(26):
            common_character_counts[i] = min(common_character_counts[i], current_character_counts[i])
    result = []
    for i in range(26):
        for _ in range(common_character_counts[i]):
            result.append(chr(i + ord('a')))
    return result


if __name__ == '__main__':
    print(commonChars(["bella", "label", "roller"]))
    print(commonChars(["cool", "lock", "cook"]))
    print(commonChars(["a", "b", "c"]))
    print(commonChars(["a"]))
    print(commonChars(["a", "a"]))
    print(commonChars(["a", "b"]))
    print(commonChars(["a", "b", "a"]))
    print(commonChars(["a", "b", "a", "b"]))
    print(commonChars(["a", "b", "a", "b", "a"]))
    print(commonChars(["a", "b", "a", "b", "a", "b"]))
    print(commonChars(["a", "b", "a", "b", "a", "b", "a"]))
    print(commonChars(["a", "b", "a", "b", "a", "b", "a", "b"]))
    print(commonChars(["a", "b", "a", "b", "a", "b", "a", "b", "a"]))
    print(commonChars(["a", "b", "a", "b", "a", "b", "a", "b", "a", "b"]))
    print(commonChars(["a", "b", "a", "b", "a", "b", "a", "b", "a", "b", "a"]))
    print(commonChars(["a", "b", "a", "b", "a", "b", "a", "b", "a", "b", "a", "b"]))
    print(commonChars(["a", "b", "a", "b", "a", "b", "a", "b", "a", "b", "a", "b", "a"]))
